vignettes/v1-2020-2021-mitma-data-codebook.qmd

```diff
@@ -0,0 +1,228 @@
+
+**Getting the data**
+
+To access the data, use the `spod_get_od()` function. In this example we will use a short interval of dates:
+
+```{r}
+dates <- c(start = "2020-02-14", end = "2020-02-17")
+od_dist <- spod_get(type = "od", zones = "dist", dates = dates)
+od_muni <- spod_get(type = "od", zones = "muni", dates = dates)
+```
+
+The data for the specified dates will be automatically downloaded and cached in the `SPANISH_OD_DATA_DIR` directory. Existing files will not be re-downloaded.
+
+**Working with the data**
+
+The resulting objects `od_dist` and `od_muni` are of class `tbl_duckdb_connection`[^1]. Basically, you can treat these as regular `data.frame`s or `tibble`s. One important difference is that the data is not actually loaded into memory, because if you requested more dates, e.g. a whole month or a year, all that data would most likely not fit into your computer's memory. A `tbl_duckdb_connection` is mapped to the downloaded CSV files that are cached on disk and the data is only loaded in small chunks as needed at the time of computation. You can manipulate `od_dist` and `od_muni` using `{dplyr}` functions such as `select()`, `filter()`, `mutate()`, `group_by()`, `summarise()`, etc. In the end of any sequence of commands you will need to add `collect()` to execute the whole chain of data manipulations and load the results into memory in an R `data.frame`/`tibble` like so:
+
+[^1]: For reference: this object also has classes: `tbl_dbi` ,`tbl_sql`, `tbl_lazy` ,and `tbl` .
+
+```{r}
+library(dplyr)
+od_mean_hourly_trips_over_the_4_days <- od_dist |>
+  group_by(time_slot) |>
+  summarise(
+    mean_hourly_trips = mean(n_trips, na.rm = TRUE),
+    .groups = "drop") |> 
+  collect()
+od_mean_hourly_trips_over_the_4_days
+```
+
+```         
+# A tibble: 24 × 2
+   time_slot mean_hourly_trips
+       <int>             <dbl>
+ 1        18              21.4
+ 2        10              19.3
+ 3         2              14.8
+ 4        15              19.8
+ 5        11              19.9
+ 6        16              19.6
+ 7        22              20.9
+ 8         0              18.6
+ 9        13              21.1
+10        19              22.5
+# ℹ 14 more rows
+# ℹ Use `print(n = ...)` to see more rows
+```
+
+In this example above, we calculated mean hourly flows over the 4 days of the requested period. The full data for all 4 days was never loaded into memory all at once. Rather the available memory of the computer was used up to its maximum limit to make that calculation happen, without ever exceeding the available memory limit. This is done transparantly to the user with the help of [`DuckDB`](https://duckdb.org/){target="_blank"} (specifically, with [{duckdb} R package](https://r.duckdb.org/index.html){target="_blank"}).
```

Comment:
A couple of comments:
1) According to Figure 1, this analysis (of four days) could have been conducted with raw CSV.gz. Is that correct? If so, these different approaches depending on the period of analysis should be explained here (referencing Figure 1).
2) Shouldn't we formally cite (add in references) the DuckDB package? I do not know the conventions for citation in these kinds of documents, but probably better to cite the most relevant R packages you use. What do you think?

Reply:
1. Figure 1 is actually just 1 day and is not connected to this code. I only inserted it go give a better impression of what the od data represents (I think it does that job quite nicely). The code for this plot is actually kind of long and to reproduce it easily, I would need to write another small function to convert the data we get with `spod_get()` into the format required for the plot.

The analysis of several days is actually conducted in this code on raw CSV.gz data. It's just that instead of reading the files one by one, combining them and making the aggregation, we are doing that using DuckDB (transparent to the user), which gives is an added benefit of cleaner data types in columns, additional columns and on-the-fly conversion to English named variables and variable values. So I'm not sure which two different approaches you are talking about. Just to be clear, even when I say "conduct analysis on raw CSV.gz data", I still mean doing that analysis using DuckDB on these CSV.gz files, without creating any actual database file or converting the data into another format. So perhaps it was just a misunderstanding.
2. Yep, we can cite, I will do it. Good idea. However, note that we also have metadata files: https://github.com/Robinlovelace/spanishoddata/blob/v1-full-support-based-on-district-level-data-only/codemeta.json and https://github.com/Robinlovelace/spanishoddata/blob/v1-full-support-based-on-district-level-data-only/CITATION.cff where all the packages we use in our package are cited.

Reply:
1. I think you misunderstood this. I meant Figure 1. I believe you are referring to Figure 2 here.
2. Perfect👌. You know more than myself about citing packages. My understanding was that if a specific package is very important for a project it is better to cite it.

Reply:
1. I did misunderstand you about which figure we are discussing. But my reply is still kind of relevant. So yes, the analysis for four days could be done on raw CSV.gz and it is.